util/src/lib.rs

```diff
@@ -80,57 +86,123 @@ fn reverse_index_bits_large<T: Copy>(arr: &[T], n_power: usize) -> Vec<T> {
     result
 }
 
-pub fn reverse_index_bits_in_place<T>(arr: &mut Vec<T>) {
-    let n = arr.len();
-    let n_power = log2_strict(n);
-
-    if n_power <= 6 {
-        reverse_index_bits_in_place_small(arr, n_power);
+/// Bit-reverse the order of elements in `arr`.
+/// SAFETY: ensure that `arr.len() == 1 << lb_n`.
+#[cfg(not(target_arch = "aarch64"))]
+unsafe fn reverse_index_bits_in_place_small<T>(arr: &mut [T], lb_n: usize) {
+    if lb_n <= 6 {
+        // BIT_REVERSE_6BIT holds 6-bit reverses. This shift makes them lb_n-bit reverses.
+        let dst_shr_amt = 6 - lb_n;
+        for src in 0..arr.len() {
+            let dst = (BIT_REVERSE_6BIT[src] as usize) >> dst_shr_amt;
+            if src < dst {
+                swap(arr.get_unchecked_mut(src), arr.get_unchecked_mut(dst));
+            }
+        }
     } else {
-        reverse_index_bits_in_place_large(arr, n_power);
+        // LLVM does not know that it does not need to reverse src at each iteration (which is
+        // expensive on x86). We take advantage of the fact that the low bits of dst change rarely and the high
+        // bits of dst are dependent only on the low bits of src.
+        let dst_lo_shr_amt = 64 - (lb_n - 6);
+        let dst_hi_shl_amt = lb_n - 6;
+        for src_chunk in 0..(arr.len() >> 6) {
+            let src_hi = src_chunk << 6;
+            let dst_lo = src_chunk.reverse_bits() >> dst_lo_shr_amt;
+            for src_lo in 0..(1 << 6) {
+                let dst_hi = (BIT_REVERSE_6BIT[src_lo] as usize) << dst_hi_shl_amt;
+                let src = src_hi + src_lo;
+                let dst = dst_hi + dst_lo;
+                if src < dst {
+                    swap(arr.get_unchecked_mut(src), arr.get_unchecked_mut(dst));
+                }
+            }
+        }
     }
 }
 
-/* Both functions below are semantically equivalent to:
-        for src in 0..n {
-            let dst = reverse_bits(src, n_power);
-            if src < dst {
-                arr.swap(src, dst);
-            }
+/// Bit-reverse the order of elements in `arr`.
+/// SAFETY: ensure that `arr.len() == 1 << lb_n`.
+#[cfg(target_arch = "aarch64")]
+unsafe fn reverse_index_bits_in_place_small<T>(arr: &mut [T], lb_n: usize) {
```

Comment:
This function panics when `lb_n=0` in debug mode, due to a shr overflow. It works in release mode though, so maybe we could just add a `debug_assert`.